### PR TITLE
PKG-4959 Pkg 3612 v2.3.0 initial release skip CI

### PR DESCRIPTION
**Pull request by danpetry**
**Destination channel:** defaults

### Links

- [PKG-4959](https://anaconda.atlassian.net/browse/PKG-4959) 
- [Upstream repository](https://github.com/triton-lang/triton)

### Explanation of changes:

- Release of triton 2.3.0 to support [pytorch v2.3.0 CUDA](https://github.com/AnacondaRecipes/pytorch-feedstock/pull/39) variant's torch.compile feature
- Update to v2.3.0
- Called "torchtriton" because it's not really in a state where it can be called "triton" and used in a wider setting. More details in the recipe.
- Many of the changes in this PR are unused changes to un-vendor llvm, mlir and CUDA tools. These were left in the recipe, but unused, with extensive comments in the recipe, because:
  - llvm/mlir: the vendored-in version is a commit half way up the main branch. So we can't release it as a package. llvm-project breaks their API a lot between versions so there isn't a compatible release.
  - CUDA: Pytorch is not compatible with CUDA v12.4 yet and has been built against CUDA v11.8. In triton, CUDA v12.1 is vendored-in, as a runtime compile tool. It's left vendored-in rather than unvendoring and using CUDA 12.4. This mirrors upstream.

### Logs

This needs to be built and tested manually, logs below. It's got a circular (test only) dependency on pytorch hence two logs.
[triton-initial-build.log](https://github.com/user-attachments/files/16059506/triton-initial-build.log)
[triton-initial-test.log](https://github.com/user-attachments/files/16059507/triton-initial-test.log)

[PKG-4959]: https://anaconda.atlassian.net/browse/PKG-4959?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: recipe/meta.yaml [hunk 1]
```diff
@@ -1,50 +1,94 @@
...
-  skip: true  # [win or cuda_compiler_version == "None"]
-  string: cuda{{ cuda_compiler_version | replace('.', '') }}py{{ CONDA_PY }}h{{ PKG_HASH }}_{{ PKG_BUILDNUM }}
+  number: 0
+  # Triton only currently supports linux, and is a GPU optimization tool.
+  # We only have a linux-64 GPU builder at the moment.
+  # It's primarily for PyTorch, and they only use it for linux-64/GPU.
+  skip: true  # [not (linux and x86_64)]
+  # the torch.compile feature in PyTorch isn't supported on python 3.12:
+  # https://github.com/pytorch/pytorch/blob/97ff6cfd9c86c5c09d7ce775ab64ec5c99230f5d/test/test_transformers.py#L3418
+  skip: true  # [py==312]
```
Comment (JeanChristopheMorinPerso) [inline recipe/meta.yaml]:
```suggestion
  skip: true  # [py>=312]
```

File: recipe/meta.yaml [hunk 2]
```diff
@@ -1,50 +1,94 @@
...
-    - mlir
-    - llvmdev
+    - ninja
+    {% if torch_commit_pin %}
+    - git
+    {% endif %}
   host:
     - python
     - pybind11
     - pip
```
Comment (cbouss) [inline recipe/meta.yaml:70]:
```suggestion
    - pip
    - setuptools
    - wheel
```
https://github.com/triton-lang/triton/blob/3f8d91bb17f6e7bc33dc995ae0860db89d351c7b/python/pyproject.toml#L3

File: recipe/meta.yaml (hunk 1 above)
Comment (danpetry) [reply recipe/meta.yaml]: Sure, this will change with pytorch 2.4.0 fyi

File: recipe/meta.yaml [hunk 3]
```diff
@@ -1,50 +1,94 @@
...
-    - llvm
-    - libmlir
-    - zlib
   run:
     - python
     - filelock
-    - pytorch =*=cuda*
-    - lit
+    - numpy
```
Comment (cbouss) [inline recipe/meta.yaml]: It doesn't show numpy here:
https://github.com/triton-lang/triton/blob/3f8d91bb17f6e7bc33dc995ae0860db89d351c7b/python/setup.py#L366

File: recipe/meta.yaml [hunk 4]
```diff
@@ -1,50 +1,94 @@
...
+    # Currently we keep CUDAtookit vendored-in (see patches section for reasoning), but this should be un-commented
+    # when this changes.
+    #
+    #- {{ compiler('cuda') }}
+    #- cuda-cuobjdump
+    #
+    #  gcc is required whether the cuda tools are vendored-in or not, to support CUDA compilation. (nvcc is called via
+    #  gcc when it processes .cu files, which are c++ syntax extended by CUDA syntax).
+    - {{ compiler('cxx') }}
+    - zlib
```
Comment (cbouss) [inline recipe/meta.yaml:89]:
```suggestion
    - zlib
  run_constrained:
    - cmake >=3.20
```
https://github.com/triton-lang/triton/blob/3f8d91bb17f6e7bc33dc995ae0860db89d351c7b/python/setup.py#L388C14-L388C25

**Review by cbouss — COMMENTED** [review]
Thanks for all the comments

File: recipe/meta.yaml (hunk 4 above)
Comment (JeanChristopheMorinPerso) [reply recipe/meta.yaml:89]: Umm, I don't think that's needed. I don't think cmake is used at runtime.

Comment (JeanChristopheMorinPerso) [reply recipe/meta.yaml:89]: Unless their cmake config files are only compatible with cmake 2.23+...

File: recipe/meta.yaml (hunk 3 above)
Comment (danpetry) [reply recipe/meta.yaml]: removed

File: recipe/meta.yaml (hunk 4 above)
Comment (danpetry) [reply recipe/meta.yaml:89]: I don't think this is needed either, it looks like it supports lit during testing: https://github.com/triton-lang/triton/commit/7c7b769e379132492d2c751e9dd68c4212bc433e
I'm pretty sure it doesn't need it to build stuff during runtime

**Comment by danpetry** [comment]
@JeanChristopheMorinPerso @cbouss are you satisfied here?

**Review by JeanChristopheMorinPerso — APPROVED** [review]

**Review by cbouss — APPROVED** [review]

**Pull request merged**